### PR TITLE
(maint) Implement equality operator

### PR DESCRIPTION
**Pull request by MikaelSmith**
Implement the equality operator for config_value classes.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
@HAIL9000 @whopper

**Comment by whopper** [comment]
@MikaelSmith woo! :+1:

**Comment by MikaelSmith** [comment]
For some reason I got distracted thinking we couldn't use a virtual `operator==`, realized last night I had no idea why I thought that.

File: lib/src/config_concatenation.cc [hunk 1]
```diff
@@ -161,9 +161,9 @@ namespace hocon {
         return make_shared<config_concatenation>(origin(), move(new_pieces));
     }
 
-    bool config_concatenation::operator==(config_concatenation const& other) const {
+    bool config_concatenation::operator==(config_value const& other) const {
         // note that "origin" is deliberately NOT part of equality
-        return (_pieces == other._pieces);
+        return equals<config_concatenation>(other, [&](config_concatenation const& o) { return _pieces == o._pieces; });
```
Comment (MikaelSmith) [inline lib/src/config_concatenation.cc]: In C++14 I'd use `[&](auto &o) { ... }` to simplify it further.

**Pull request merged**